post/2023/beyond-sf/index.qmd

```diff
@@ -0,0 +1,305 @@
+  )
+
+y <- dmnd_sf |> 
+  group_by(clarity) |> 
+  summarise(avg_price = mean(price))
+
+all.equal(x, y, check.attributes = FALSE) # <1>
+```
+1. We ignore attributes because they _will_ be different because the classes are different
+
+## Other geometry vectors
+
+Knowing how `sf` works can be helpful for understanding how we can ease our reliance on it for all geospatial operations. Instead of thinking of the entire sf data frame as the thing that handles all geometry operations we now know that it is the `sfc` geometry column. 
+
+In R there are different ways of representing geometry besides `sfc` vectors. The packages `s2`, `geos`, `wk`, and `rsgeo` all provide different vectors of geometries that can be used. 
+
+These libraries are very handy for doing geometric operations. Each of these packages tend to be better at one thing than another and each have their place. You will often find big speed improvements if you use these libraries instead of `sf` for certain things.
+
+Take for example calculating the length of linestrings on a geodesic. We can use the `roxel` dataset from `{sfnetworks}`. 
+
+```{r}
+data(roxel, package = "sfnetworks")
+```
+
+To illustrate we can extract the geometry column and cast it as an `rsgeo` class object.
+
+```{r}
+library(rsgeo)
+geo <- roxel$geometry
+rs <- as_rsgeo(geo)
+```
+
+We can then use the functions `st_length()` and `length_haversine()` respectively to compute the length of linestrings. 
+
+```{r}
+bench::mark(
+  st_length(geo),
+  length_haversine(rs),
+  check = FALSE
+)
+```
+This is markedly faster when using rsgeo. We also do not have to extract the vector and work with it as its own object. Any vector can be included in a data frame, remember? 
+
+```{r}
+roxel |> 
+  as_tibble() |> 
+  mutate(
+    geometry = as_rsgeo(geometry),
+    length = length_haversine(geometry)
+    ) |> 
+  select(name, type, length, geometry)
+```
+This same approach can work for other libraries such as geos.
+
+```{r}
+library(geos)
+
+roxel |> 
+  as_tibble() |> 
+  mutate(
+    geometry = as_geos_geometry(geometry),
+    length = geos_length(geometry)
+    ) |> 
+  select(name, type, length, geometry)
+```
+
+If you plan on doing more than just one operation, it is best to convert to the desired geometry type only once and then use it in subsequent calls. Converting from one geometry type can take some time and add additional and unwanted overhead if you're doing it all the time. 
+
+```{r}
+roxy_geos <- roxel |> 
+  st_transform(25832) |> # <1>
+  as_tibble() |> # <2>
+  mutate(geometry = as_geos_geometry(geometry)) # <3>
+```
+1. We project our dataset to use meters
+2. We remove the sf class so we can modify the geometry column
+3. We update the geometry from an `sfc` to a `geos_geometry`
+
+With a geometry column that is a `geos_geometry`, you can use the functions from the `geos` package to perform your analyses. And, if doing larger scale analyses, it might be faster and more memory efficient!
```

Comment:
```suggestion
With a geometry column that is a `geos_geometry`, you can use the functions from the **geos** package to perform your analyses. And, if doing larger scale analyses, it might be faster and more memory efficient!
```